deepspeed/inference/v2/ragged/prefix_block_map.py

```diff
@@ -0,0 +1,56 @@
+# Copyright (c) Microsoft Corporation.
+# SPDX-License-Identifier: Apache-2.0
+
+# DeepSpeed Team
+
+from typing import Dict, List, Set
+import hashlib
+
+import torch
+
+
+def token_ids_to_hash(token_ids: torch.Tensor):
+    # Convert the tensor to bytes
+    tensor_bytes = token_ids.numpy().tobytes()
+    hash_obj = hashlib.sha256()
+    # Update the hash object with the bytes
+    hash_obj.update(tensor_bytes)
+    # Get the hexadecimal digest of the hash
+    return hash_obj.hexdigest()
+
+
+class PrefixBlockMap():
+
+    def __init__(self, block_size: int):
+        self.tokens_to_blocks: Dict[str, List[int]] = {}
+        self.blocks_to_tokens: Dict[Set[int], str] = {}
+        self.block_size: int = block_size
+
+    def lookup(self, tokens: torch.Tensor) -> torch.Tensor:
+        n_blocks = len(tokens) // self.block_size
+        cached_blocks = torch.tensor([], dtype=torch.int32)
+        for i in range(n_blocks):
+            chunk = tokens[:(i + 1) * self.block_size]
+            hash = token_ids_to_hash(chunk)
+            if hash in self.tokens_to_blocks:
+                cached_blocks = self.tokens_to_blocks[hash]
+            else:
+                break
+        return cached_blocks
+
+    def extend(self, tokens: torch.Tensor, new_block_ids: List[int], num_already_cached_blocks: int) -> None:
+        n_blocks = len(tokens) // self.block_size
+        for i in range(num_already_cached_blocks, n_blocks):
+            chunk = tokens[:(i + 1) * self.block_size]
+            hash = token_ids_to_hash(chunk)
+            if hash not in self.tokens_to_blocks:
```

Comment:
Do we still need this if statement if in the for loop you are already starting from num_already_cached_blocks.

Reply:
I'm not sure if it is necessary but there might be a complicated case. Assume we are running two requests.
Request 1 is using cached blocks 0, 1, 2, and just generated the last token of the current block. Then it saves the generated sequence and a hash.
Request 2 is also using the cached block 0, 1, 2, but the generation is a few steps later than Request 1. They are not sharing the last block. But the request may generate exact same tokens for the last block. So, it will try to update the cache with the same hash.
In this case, I didn't want to overwrite the cache.

Reply:
sounds good, so this is to ensure we are sharing the recently generated blocks.